### DISCUSSION TITLE
Problem with setting up MFA in my Next.js Application.

### DISCUSSION THREAD

**Discussion by ishaanbedi** [supabase/supabase#10246] [Questions] [answered]
I got early access to MFA, and while following the docs for the same and setting it up in my Next.js app, I'm facing a problem:

<img width="764" alt="Screenshot 2022-11-10 at 9 32 02 PM" src="https://user-images.githubusercontent.com/39641326/201144854-fef1490b-afa1-4f99-ad8d-ccd3eabd8c54.png">

The function that I've written which gets triggered after verifying that the user is logged in and Supabase session is active, goes like this: 
```
  const enableTwoFA = async () => {
    setLoading(true);
    const { data, error } = await supabase.auth.mfa.enroll({
      factorType: "totp",
    });
    if (error) {
      console.log(error);
    }
    if (data) {
      console.log(data);
    }
    setLoading(false);
  };
```
Looking forward to someone helping me out with this.
@J0

**Comment by J0** [comment, marked as answer]
Hey @ishaanbedi,

Thanks for typing out the query. The error arises as there was a misconfiguration on the server which caused the feature flag on MFA to be toggled to false anytime another GoTrue related dashboard settings is configured. As such, the above error is happening because of the misconfigured feature flag. I've put out a fix for this issue but for now I've reset your instance so that you can use MFA again.


Sorry for the confusion caused and please let me know if you face any further issues.

Thanks!
Joel

**Reply by ishaanbedi** [reply]
Alright, @J0 !
Thank you for clarifying it. 
It's working perfectly fine now!

**Comment by ishaanbedi** [comment]
Hi @J0 
<img width="1065" alt="Screenshot 2022-11-11 at 9 39 28 PM" src="https://user-images.githubusercontent.com/39641326/201382545-6209c792-bd6a-4441-9b08-5ece87b09949.png">
I can use this feature now, but after around 7-8 API calls to `supabase.auth.mfa.enroll({})`, I'm getting the error as shown in the image. 
Here's the code:
```
  const onEnableClicked = () => {
    console.log("here");
    (async () => {
      const challenge = await supabase.auth.mfa.challenge({ factorID });
      if (challenge.error) {
        console.log(challenge.error.message);
        throw challenge.error;
      }
      const challengeId = challenge.data.id;
      const verify = await supabase.auth.mfa.verify({
        factorID,
        challengeId,
        code: verifyCode,
      });
      if (verify.error) {
        console.log(verify.error.message);
        throw verify.error;
      }
      alert("You have successfully enabled TFA");
    })();
  };

  useEffect(() => {
    (async () => {
      const { data, error } = await supabase.auth.mfa.enroll({
        factorType: "totp",
      });
      if (error) {
        throw error;
      }
      setFactorID(data.id);
      setQRSvg(data.totp.qr_code);
      console.log(data.totp);
    })();
  }, []);
```

I'm unable to understand how to unenroll the user. Is it a problem with the implementation or something else?

**Reply by J0** [reply]
Hey @ishaanbedi,

Thanks for the query! The error arises as by default each user can only have a maximum of 10 unverified unenrolled factors - this is to ensure that users do not flood the database by enrolling. 

To unenroll the factor, you can use the [`unenroll` endpoint](https://github.com/supabase/gotrue-js/blob/master/src/GoTrueClient.ts#L1138) on the SDK:

```js
const { data, error } = await supabase.auth.mfa.unenroll({
        factorId: <your_id>,
      });
```




Joel

**Reply by ishaanbedi** [reply]
Oh, okay!
Thank you for the clarification @J0 .
I should be careful while using the useEffect hook, which is causing API calls on each render while working with the UI.

**Reply by romanroibu** [reply]
Hey @J0, I'm currently trying to solve the "maximum of 10 unverified unenrolled factors" issue in a graceful way. I came up with this solution:

https://github.com/romanroibu/megabase/blob/main/migrations/20240213093944_remove_other_unverified_mfa_factors.sql

Would appreciate if you could comment on it - are there any downsides to doing things this way? Any potential problem that I might be overlooking here?

Thanks!
